src_kt/java/arcs/storage/util/ProxyCallbackManager.kt

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2019 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.storage.util
+
+import arcs.crdt.CrdtData
+import arcs.crdt.CrdtOperation
+import arcs.storage.ProxyCallback
+import arcs.storage.ProxyMessage
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/** Thread-safe manager of a collection of [ProxyCallback]s. */
+class ProxyCallbackManager<Data : CrdtData, Op : CrdtOperation, ConsumerData> {
+    private val mutex = Mutex()
+    internal var nextCallbackToken = 1
+    internal val callbacks = mutableMapOf<Int, ProxyCallback<Data, Op, ConsumerData>>()
+
+    /** Adds a [ProxyCallback] to the collection, and returns its token. */
+    fun register(proxyCallback: ProxyCallback<Data, Op, ConsumerData>): Int {
+        while (!mutex.tryLock()) { /* Wait. */ }
```

Comment:
Is this really the best way of doing this? Does Kotlin have a thread-safe map implementation you could use instead?

Reply:
It doesn't have one without the JVM, unfortunately... I think another way to do a lot of this thread safety would be to use Actors, but I don't feel comfortable with them just yet to make that call-- and I think most engineers understand mutexing.

Reply:
Ah of course. Thanks for explaining! Yeah the mutex is fine, I just thought there were safer/more sophisticated tools available, but I guess not in kotlin native.